kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemDataset.java

```diff
@@ -199,11 +199,14 @@ public PathIterator pathIterator() {
     try {
       if (!fileSystem.exists(partitionDirectory)) {
         if (allowCreate) {
-          fileSystem.mkdirs(partitionDirectory);
           if (partitionListener != null) {
             partitionListener.partitionAdded(namespace, name,
                 toRelativeDirectory(key).toString());
           }
+
+          // ensure that the directory exists, it may or may not have been
+          // created by the partitionListener
+          fileSystem.mkdirs(partitionDirectory);
```

Comment:
Here's the Java doc:

http://hadoop.apache.org/docs/r2.5.2/api/org/apache/hadoop/fs/FileSystem.html#mkdirs(org.apache.hadoop.fs.Path,%20org.apache.hadoop.fs.permission.FsPermission)